llvm/lib/Analysis/InstructionSimplify.cpp

```diff
@@ -5080,6 +5237,49 @@ static Value *simplifyUnaryIntrinsic(Function *F, Value *Op0,
     if (llvm::cheri::isKnownUntaggedCapability(Op0, &Q.DL))
       return Constant::getNullValue(F->getReturnType());
     break;
+  case Intrinsic::cheri_cap_sealed_get:
+  case Intrinsic::cheri_cap_flags_get:
+  case Intrinsic::cheri_cap_base_get:
+    // Values derived from NULL and where the only operations that have been
+    // applied are address manipulations, always have the following properties:
+    // - base is zero
+    // - unsealed
+    // - permissions are zero
+    // - flags are zero
+    // However some fields may vary across architectures:
+    // - length either -1/0
+    // - type either -1/0
+    // TODO: for some of these we could ignore more things
+    if (isa<ConstantPointerNull>(
+            getBasePtrIgnoringCapabilityAddressManipulation(Op0, Q.DL))) {
+      return Constant::getNullValue(F->getReturnType());
+    }
+    break;
+  case Intrinsic::cheri_cap_perms_get:
+    // TODO: For getperms we can ignore almost all CHERI intrinsics since there
+    // is no setperms that could add permission bits.
+    // However, this is unlikely to be particularly beneficial to optimizations,
+    // so be conservative for now.
+    if (isa<ConstantPointerNull>(
+            // Maybe: getBasePtrIgnoringCapabilityManipulation(Op0, Q.DL))) {
+            getBasePtrIgnoringCapabilityAddressManipulation(Op0, Q.DL))) {
+      return Constant::getNullValue(F->getReturnType());
+    }
+    break;
+    // Note: No optimizations for getting type/length optimizing this since this
+    // could differ across implementations of CHERI.
```

Comment:
```suggestion
  // Note: No optimizations for getting type/length optimizing this since this
  // could differ across implementations of CHERI.
```